servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/ConnectionPoolStrategy.java

```diff
@@ -0,0 +1,158 @@
+
+import io.servicetalk.client.api.LoadBalancedConnection;
+import io.servicetalk.context.api.ContextMap;
+
+import java.util.List;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.function.Predicate;
+import javax.annotation.Nullable;
+
+import static io.servicetalk.utils.internal.NumberUtils.ensureNonNegative;
+import static java.lang.Math.min;
+
+interface ConnectionPoolStrategy<C extends LoadBalancedConnection> {
+
+    /**
+     * With a relatively small number of connections we can minimize connection creation under moderate concurrency by
+     * exhausting the full search space without sacrificing too much latency caused by the cost of a CAS operation per
+     * selection attempt.
+     */
+    int MIN_RANDOM_SEARCH_SPACE = 64;
+
+    /**
+     * For larger search spaces, due to the cost of a CAS operation per selection attempt we see diminishing returns for
+     * trying to locate an available connection when most connections are in use. This increases tail latencies, thus
+     * after some number of failed attempts it appears to be more beneficial to open a new connection instead.
+     * <p>
+     * The current heuristics were chosen based on a set of benchmarks under various circumstances, low connection
+     * counts, larger connection counts, low connection churn, high connection churn.
+     */
+    float RANDOM_SEARCH_FACTOR = 0.75f;
+
+    /**
+     * Select a connection from the ordered list of connections.
+     * @param connections the list of connections to pick from.
+     * @param selector a predicate used to test a connection.
+     * @param context optional {@link ContextMap} that can be used to pass information.
+     * @return the selected connection, or {@code null} if no existing connection was selected.
+     */
+    @Nullable
+    C select(List<C> connections, Predicate<C> selector, @Nullable ContextMap context);
+
+    /**
+     * A connection selection strategy that prioritizes a configurable "core" pool size and then will
+     * spill over to an overflow pool.
+     * <p>
+     * This {@link ConnectionPoolStrategy} attempts to emulate the pooling behavior often seen in thread pools.
+     * Specifically it allows for the configuration of a "core pool" size which are intended to be long-lived.
+     * Members of this core pool are selected at random that so they will get an even traffic load and, because they
+     * are equally selectable, will tend not to be removed due to idleness.
+     * <p>
+     * If the core pool cannot satisfy the load traffic can spill over to extra connections which are selected in-order.
+     * This has the property of minimizing traffic to the latest elements added outside the core pool size, thus let
+     * them tend to idle out of the pool once they're no longer of use.
+     * @param <C> the concrete type of the {@link LoadBalancedConnection}.
+     */
+    final class CorePoolStrategy<C extends LoadBalancedConnection> implements ConnectionPoolStrategy<C> {
```

Comment:
Consider moving each strategy to its independent class. If necessary to provide public access, we can either use a builder or static factories pattern we use in other places, like `ConnectionPoolStrategies` class with static methods to build different strategies.

Reply:
Agreed, I'll move them. I left them here in the draft just so I didn't invest too much time if the idea was outright rejected.